### ISSUE TITLE
Unused variables - false positives

### ISSUE THREAD

**Issue by movedoa** [Riverside-Software/sonar-openedge#25]
V 1.9.2

1. I have a class with a global private variable.
This variable is set in the constructor of the class with an input paramter of the constructor.
But the unused variable rule reports that this variable is never used.

2. Nearly every OUTPUT variable is a false positive since most of the time they only get set and are never used in the procedure or method
OUTPUT Variables should only be marked as unused when they are never set. (the same for INPUT-OUTPUT)

3. Nearly Every .net Event Method has falso positives for its input parameters (sender and e), there is no way to avoid the variables that the event wants, so parameters of event methods should be ignored

**Comment by movedoa** [comment]
The same thing seems to happen with output parametrs in procedures

**Comment by gquerret** [comment]
Part 1 : what do you call a global private variable ?
Part 2 : that's for rule 'Variable is assigned but never read', right ?
Part 3 : agreed

**Comment by movedoa** [comment]
1.) I am not shure what you mean but i the bug happens with class scope variables
2.) Oh sry yes, i mixed these two in my head (still a false positive though)

**Comment by jakejustus** [comment]
1. So just something like:
```
class rssw.foo:
 define private variable bar as char.
 public constructor rssw.foo(prm as char):
  this-object:bar = prm.
 end constructor.
end class.
```
raise a false positive ? Or something different ?

2. Thanks, wanted to make sure

**Comment by movedoa** [comment]
yeah pretty much, an example would be a class like this:

class rssw.foo:
 define private variable bar as char.
 public constructor rssw.foo(prm as char):
  this-object:bar = prm.
 end constructor.

 method public void foo():
    bar:foomethod();
 end method.

end class.

**Comment by gquerret** [comment]
Sorry, but not able to reproduce 1. using this example. Could you refine the example, please ?
Point 2. has been fixed.

**Comment by movedoa** [comment]
I tried it with the example and i can't reproduce it either.
Nevertheless the problem exists in my code.
I could send you the file but i don't think it would help because you wouldn't be able to compile them :/

**Comment by movedoa** [comment]
Sill getting "Variable is assigned but never read" for OUTPUT variables in 1.9.4

**Comment by movedoa** [comment]
Ok i could reproduce error 1.

> Unused variable bar

> The content of a variable is neither read nor written during the flow of execution. Those variables should be removed. 

```
CLASS SonarTest:

   DEFINE PRIVATE STATIC VARIABLE bar AS CHARACTER NO-UNDO.

   CONSTRUCTOR SonarTest(prm AS CHARACTER):
      SonarTest:bar = prm.
   END CONSTRUCTOR.

   METHOD PUBLIC VOID foo():
      THIS-OBJECT:foo2(SonarTest:bar).
   END METHOD.

   METHOD PUBLIC VOID foo2(
      pcBar AS CHARACTER
   ):
      /* Do something with pcBar.
         Irl, this mehthod is in some other class (I don't know if that makes any difference */
   END METHOD.
END CLASS.
```

**Comment by gquerret** [comment]
OUTPUT variables shouldn't be reported anymore as of version 1.0.7 of riverside-rules plugin

**Comment by movedoa** [comment]
You are right, my bad, just looked again and couldn't find any.
I guess i was in an old sonar project by accident.

**Comment by gquerret** [comment]
Created issues #62 and #63, and closing this issue
